tools/autograd/gen_variable_type.py

```diff
@@ -273,6 +274,51 @@
 #endif
 """)
 
+FW_DERIVATIVE_CHECK_TEMPLATE = CodeTemplate("""\
+isFwGradDefined(${req_inp})\
+""")
+
+FW_DERIVATIVE_DEFINED_GRAD_TEMPLATE = CodeTemplate("""\
+auto ${inp}_fw_grad = toLegacyFwGrad(${inp});
```

Comment:
This is entirely new code, I'm not sure why there is a "Legacy" in the name of the function it's calling

Reply:
This is to make the link between the legacy autograd API that takes plain Tensors (that can be undefined) and the codegen that now sometimes give optional Tensors.